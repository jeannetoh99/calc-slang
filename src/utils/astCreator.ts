import * as es from '../ast'

export const getValueDeclarationName = (decl: es.ValueDeclaration) =>
  (decl.declarations[0].id as es.Identifier).name

export const locationDummyNode = (line: number, column: number) =>
  literal('Dummy', 'string', { start: { line, column }, end: { line, column } })

export const identifier = (name: string, loc?: es.SourceLocation | null): es.Identifier => ({
  type: 'Identifier',
  name,
  loc
})

export const literal = (
  value: any,
  litType: es.Type,
  loc?: es.SourceLocation | null
): es.Literal => ({
  type: 'Literal',
  value,
  litType,
  loc
})

export const expressionStatement = (expression: es.Expression): es.ExpressionStatement => ({
  type: 'ExpressionStatement',
  expression
})

export const program = (body: es.Statement[]): es.Program => ({
  type: 'Program',
  sourceType: 'module',
  body
})
