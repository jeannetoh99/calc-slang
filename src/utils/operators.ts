import { Thunk } from '../types'

export function forceIt(val: Thunk | any): any {
  if (val !== undefined && val !== null && val.isMemoized !== undefined) {
    if (val.isMemoized) {
      return val.memoizedValue
    }

    const evaluatedValue = forceIt(val.f())

    val.isMemoized = true
    val.memoizedValue = evaluatedValue

    return evaluatedValue
  } else {
    return val
  }
}
