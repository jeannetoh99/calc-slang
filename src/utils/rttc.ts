import * as es from '../ast'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import { ErrorSeverity, ErrorType } from '../types'

export class TypeError extends RuntimeSourceError {
  public type = ErrorType.RUNTIME
  public severity = ErrorSeverity.ERROR
  public location: es.SourceLocation

  constructor(node: es.Node, public side: string, public expected: string, public got: string) {
    super(node)
  }

  public explain() {
    return `Expected ${this.expected}${this.side}, got ${this.got}.`
  }

  public elaborate() {
    return this.explain()
  }
}

export const isInt = (v: es.Literal) => v.smlType.type === 'int'
export const isReal = (v: es.Literal) => v.smlType.type === 'real'
export const isNum = (v: es.Literal) => isInt(v) || isReal(v)
export const isString = (v: es.Literal) => v.smlType.type === 'string'
export const isBool = (v: es.Literal) => v.smlType.type === 'bool'
export const isUnit = (v: es.Literal) => v.smlType.type === 'unit'

export const checkIsInt = (node: es.Node, side: string, test: es.Literal) => {
  return isInt(test) ? undefined : new TypeError(node, side, 'int', test.smlType.type)
}

export const checkIsReal = (node: es.Node, side: string, test: es.Literal) => {
  return isReal(test) ? undefined : new TypeError(node, side, 'real', test.smlType.type)
}

export const checkIsNum = (node: es.Node, side: string, test: es.Literal) => {
  return isNum(test) ? undefined : new TypeError(node, side, 'real or int', test.smlType.type)
}

export const checkIsString = (node: es.Node, side: string, test: es.Literal) => {
  return isString(test) ? undefined : new TypeError(node, side, 'string', test.smlType.type)
}

export const checkIsBool = (node: es.Node, side: string, test: es.Literal) => {
  return isBool(test) ? undefined : new TypeError(node, side, 'boolean', test.smlType.type)
}

export const checkIsUnit = (node: es.Node, side: string, test: es.Literal) => {
  return isUnit(test) ? undefined : new TypeError(node, side, 'unit', test.smlType.type)
}

export const checkIsList = (node: es.Node, side: string, test: es.Literal) => {
  return isUnit(test) ? undefined : new TypeError(node, side, 'unit', test.smlType.type)
}


export const checkIsType = (node: es.Node, side: string, test: es.Literal, type: es.LiteralType) => {
  switch (type.type) {
    case 'int':
      return checkIsInt(node, side, test)
    case 'real':
      return checkIsReal(node, side, test)
    case 'string':
      return checkIsString(node, side, test)
    case 'bool':
      return checkIsBool(node, side, test)
    case 'unit':
      return checkIsUnit(node, side, test)
  }
}
