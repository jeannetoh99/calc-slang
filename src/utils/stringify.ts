import { isInteger } from 'lodash'

import { List, ListType, SmlValue, Tuple, TupleType, Type, VariableType } from '../ast'
import { ProgramResult, ResultType, Value } from '../types'

export interface ArrayLike {
  replPrefix: string
  replSuffix: string
  replArrayContents: () => Value[]
}

export type DecResType = {
  name: string
  value: Value
  type: string
}

const formatResult = (result: DecResType) => {
  return ['val', result.name, '=', result.value, ':', result.type].join(' ') + ';'
}

export const extractVariableTypes = (type: Type): VariableType[] => {
  switch (type.type) {
    case 'function':
      return extractVariableTypes(type.paramType).concat(extractVariableTypes(type.returnType))
    case 'list':
      return extractVariableTypes(type.elementType)
    case 'tuple':
      return type.elementTypes.flatMap(elemType => extractVariableTypes(elemType))
    case 'variable':
      return [type]
    default:
      return []
  }
}

export const stringifyVarTypes = (type: Type): string => {
  const varTypes = extractVariableTypes(type)
  let res = ''
  if (varTypes.length != 0) {
    res = '\u2200 ' + stringifyType(varTypes[0])
    for (let i = 1; i < varTypes.length; i++) {
      res += ',' + stringifyType(varTypes[i])
    }
    res += ' . '
  }
  return res
}

export const stringifyType = (type: Type, inner: boolean = false): string => {
  return type.type === 'function'
    ? (type.paramType ? stringifyType(type.paramType) : '?') +
        ' -> ' +
        (type.returnType ? stringifyType(type.returnType) : '?')
    : type.type === 'list'
    ? (type.elementType ? stringifyType(type.elementType) : '?') + ' list'
    : type.type === 'tuple'
    ? (inner ? '(' : '') +
      type.elementTypes.map(elemType => stringifyType(elemType, true)).join(' * ') +
      (inner ? ')' : '')
    : type.type === 'variable'
    ? "'v" + type.id
    : type.type
}

const stringifyValue = (value: SmlValue, type: Type): Value => {
  switch (type.type) {
    case 'string': {
      return '"' + value.value + '"'
    }
    case 'unit': {
      return '()'
    }
    case 'real': {
      return value.value + (isInteger(value.value) ? '.0' : '')
    }
    case 'list': {
      const list = value as List
      return '[' + list.value.map(val => stringifyValue(val, type.elementType)) + ']'
    }
    case 'tuple': {
      const tuple = value as Tuple
      const elementStrings = []
      for (let i = 0; i < tuple.value.length; i++) {
        elementStrings.push(stringifyValue(tuple.value[i], type.elementTypes[i]))
      }
      return '(' + elementStrings + ')'
    }
    default: {
      return value.value
    }
  }
}

const extractMatch = (name: string, value: Value, type: Type): DecResType => {
  return {
    name,
    value: stringifyValue(value, type),
    type: stringifyVarTypes(type) + stringifyType(type)
  }
}

const extractDeclaration = (res: ResultType, type: Type): DecResType[] => {
  console.log(res, type)
  if (res.pat.type === 'Identifier') {
    return [extractMatch(res.pat.name, res.value, type)]
  }

  if (res.pat.type === 'TuplePattern') {
    const matches: DecResType[] = []
    const elemTypes = (type as TupleType).elementTypes
    const patterns = res.pat.elements
    const values = (res.value as Tuple).value
    for (let i = 0; i < res.pat.elements.length; i++) {
      const pat = patterns[i]
      if (pat.type === 'Identifier') {
        matches.push(extractMatch(pat.name, values[i], elemTypes[i]))
      }
    }
    return matches
  }

  // If pattern is wildcard or literal
  return []
}

export const formatResults = (result: DecResType[]) => {
  return result.map(formatResult).join('\n')
}

export const stringify = (value: Value): string => {
  const programResult = value as ProgramResult
  const res = []
  for (let i = 0; i < programResult.values.length; i++) {
    res.push(...extractDeclaration(programResult.values[i], programResult.types[i]))
  }
  return res.map(formatResult).join('\n')
}
