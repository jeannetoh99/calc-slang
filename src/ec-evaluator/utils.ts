import { uniqueId } from 'lodash'

import { Context } from '..'
import * as es from '../ast'
import * as errors from '../errors/errors'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import { arity } from '../stdlib/misc'
import { Environment, Frame, Value } from '../types'
import * as ast from '../utils/astCreator'
import { builtinFunctions, builtinInfixFunctions } from './builtin'
import * as instr from './instrCreator'
import { Agenda } from './interpreter'
import { AgendaItem, BuiltinInstr, ClosureInstr, Instr, InstrType } from './types'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(...items: T[]): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  private storage: T[] = []

  public constructor() {}

  public push(...items: T[]): void {
    for (const item of items) {
      this.storage.push(item)
    }
  }

  public pop(): T | undefined {
    return this.storage.pop()
  }

  public peek(): T | undefined {
    if (this.size() === 0) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  public size(): number {
    return this.storage.length
  }
}

/**
 * Typeguard for esNode to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an esNode and false if it is an instruction.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}

/**
 * Typeguard for esIdentifier. To verify if an esNode is an esIdentifier.
 *
 * @param node an esNode
 * @returns true if node is an esIdentifier, false otherwise.
 */
export const isIdentifier = (node: es.Node): node is es.Identifier => {
  return (node as es.Identifier).name !== undefined
}

/**
 * A helper function for handling sequences of statements.
 * Statements must be pushed in reverse order, and each statement is separated by a pop
 * instruction so that only the result of the last statement remains on stash.
 * Value producing statements have an extra pop instruction.
 *
 * @param seq Array of statements.
 * @returns Array of commands to be pushed into agenda.
 */
export const handleSequence = (seq: es.Statement[]): AgendaItem[] => {
  const result: AgendaItem[] = []
  let valueProduced = false
  for (const command of seq) {
    valueProduced ? result.push(instr.popInstr()) : (valueProduced = true)
    result.push(command)
  }
  return result.reverse()
}

/**
 * Environments
 */

export const currentEnvironment = (context: Context) => context.runtime.environments[0]

export const createEnvironment = (
  closure: ClosureInstr,
  args: Value[],
  callExpression: es.CallExpression
): Environment => {
  const id = uniqueId()
  const environment: Environment = {
    name: isIdentifier(callExpression.callee)
      ? callExpression.callee.name
      : 'anonymous closure ' + id,
    tail: closure.env,
    head: {},
    id,
    callExpression: {
      ...callExpression,
      args
    }
  }
  closure.srcNode.params.forEach((param, index) => {
    environment.head[(param as es.Identifier).name] = args[index]
  })
  return environment
}

export const popEnvironment = (context: Context) => context.runtime.environments.shift()

export const pushEnvironment = (context: Context, environment: Environment) => {
  context.runtime.environments.unshift(environment)
  context.runtime.environmentTree.insert(environment)
}

export const createBlockEnvironment = (
  context: Context,
  name = 'blockEnvironment',
  head: Frame = {}
): Environment => {
  return {
    name,
    tail: currentEnvironment(context),
    head,
    id: uniqueId()
  }
}
/**
 * Variables
 */

const DECLARED_BUT_NOT_YET_ASSIGNED = Symbol('Used to implement hoisting')

function declareIdentifier(context: Context, name: string, node: es.Node) {
  const environment = currentEnvironment(context)
  if (!environment.head.hasOwnProperty(name)) {
    environment.head[name] = DECLARED_BUT_NOT_YET_ASSIGNED
  }
  return environment
}

function declareVariables(context: Context, node: es.ValueDeclaration) {
  for (const declaration of node.declarations) {
    declareIdentifier(context, (declaration.id as es.Identifier).name, node)
  }
}

export function declareFunctionsAndVariables(context: Context, node: es.BlockStatement) {
  for (const statement of node.body) {
    switch (statement.type) {
      case 'ValueDeclaration':
        declareVariables(context, statement)
        break
      case 'FunctionDeclaration':
        declareIdentifier(context, (statement.id as es.Identifier).name, statement)
        break
    }
  }
}

export function defineVariable(context: Context, name: string, value: Value, constant = false) {
  const environment = currentEnvironment(context)

  Object.defineProperty(environment.head, name, {
    value,
    writable: !constant,
    enumerable: true
  })

  return environment
}

export const getVariable = (context: Context, name: string, node: es.Identifier) => {
  let environment: Environment | null = currentEnvironment(context)
  while (environment) {
    if (environment.head.hasOwnProperty(name)) {
      if (environment.head[name] === DECLARED_BUT_NOT_YET_ASSIGNED) {
        return handleRuntimeError(context, new errors.UnassignedVariable(name, node))
      } else {
        return environment.head[name]
      }
    } else {
      environment = environment.tail
    }
  }
  return handleRuntimeError(context, new errors.UndefinedVariable(name, node))
}

export const handleRuntimeError = (context: Context, error: RuntimeSourceError) => {
  context.errors.push(error)
  context.runtime.environments = context.runtime.environments.slice(
    -context.numberOfOuterEnvironments
  )
  throw error
}

export const checkNumberOfArguments = (
  context: Context,
  callee: ClosureInstr | BuiltinInstr,
  args: Value[],
  exp: es.CallExpression
) => {
  if (callee?.instrType === InstrType.CLOSURE) {
    // User-defined or Pre-defined functions
    const instr = callee as ClosureInstr
    const params = instr.srcNode?.params
    if (params.length !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(exp, params.length, args.length)
      )
    }
  } else if (callee?.instrType === InstrType.BUILTIN) {
    const instr = callee as BuiltinInstr
    if (instr.arity !== args.length) {
      return handleRuntimeError(
        context,
        new errors.InvalidNumberOfArguments(exp, instr.arity, args.length)
      )
    }
  }
  return undefined
}

/**
 * This function can be used to check for a stack overflow.
 * The current limit is set to be an agenda size of 1.0 x 10^5, if the agenda
 * flows beyond this limit an error is thrown.
 * This corresponds to about 10mb of space according to tests ran.
 */
export const checkStackOverFlow = (context: Context, agenda: Agenda) => {
  if (agenda.size() > 100000) {
    const stacks: es.CallExpression[] = []
    let counter = 0
    for (
      let i = 0;
      counter < errors.MaximumStackLimitExceeded.MAX_CALLS_TO_SHOW &&
      i < context.runtime.environments.length;
      i++
    ) {
      if (context.runtime.environments[i].callExpression) {
        stacks.unshift(context.runtime.environments[i].callExpression!)
        counter++
      }
    }
    handleRuntimeError(
      context,
      new errors.MaximumStackLimitExceeded(context.runtime.nodes[0], stacks)
    )
  }
}

export const populateBuiltInIdentifiers = (context: Context) => {
  for (const key in builtinFunctions) {
    const builtinInstr = instr.builtinInstr(key, arity(builtinFunctions[key]), false)
    defineVariable(context, key, builtinInstr)
  }
}
