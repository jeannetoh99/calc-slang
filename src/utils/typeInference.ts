import { cloneDeep } from 'lodash'
import { Context } from '..'
import * as es from '../ast'
import { builtinFunctionTypes } from '../ec-evaluator/builtin'
import { TypeError, TypeInferenceError } from '../errors/typeErrors'
import { boolType, functionType, listType, tupleType, variableType } from './astCreator'
import { isTypeEqual } from './rttc'

class TypeEnv {
  constructor(public map: { [name: string]: es.Type } = {}) {}

  get(name: any) {
    if (name in this.map) {
      return this.map[name]
    } else {
      return undefined
    }
  }

  extend(name: any, val: es.Type) {
    return new TypeEnv(Object.assign({}, this.map, { [name]: val }))
  }
}

export interface InferResult {
  type: es.Type
  constraints: Constraint[]
  env: TypeEnv
}


/**
 *  Infer the type of a node
 */
export function infer(node: es.TypedNode, env: TypeEnv): InferResult {
  console.log("start")
  console.log(cloneDeep(node))
  const res = addConstraints(node, env)
  console.log(res)
  const subs = unify(res.constraints)
  console.log(subs)
  node.smlType = substitute(res.type, subs)
  console.log(cloneDeep(node))
  console.log("end")
  return {
    type: node.smlType,
    constraints: [],
    env: res.env
  }
}

export function bind(pat: es.Pattern, type: es.Type, env: TypeEnv): TypeEnv {
  if (pat.type === 'TuplePattern' && type.type === 'tuple') {
    // TODO: check no duplicate identifiers in the same pat
    for (let i = 0; i < pat.elements.length; i++) {
      env = bind(pat.elements[i], type.elementTypes[i], env)
    }
    return env
  } else if (pat.type === 'Identifier') {
    return env.extend(pat.name, type)
  } else if (pat.type === 'Literal') {
    // check that literal pat type and type is the same
  }
  return env
}

export function addConstraints(node: es.Node, env: TypeEnv): InferResult {
  switch (node.type) {
    case 'Program':
    case 'BlockStatement': {
      let constraints : Constraint[] = []
      let type
      for (const stmt of node.body) {
        const res = infer(stmt, env)
        type = res.type
        env = res.env
      }
      return {
        type: type ?? node.smlType,
        constraints,
        env
      }
    }
    case 'ExpressionStatement': {
      const res = infer(node.expression, env)
      return {
        type: res.type,
        constraints: [],
        env: res.env
      }
    }
    case 'ValueDeclaration':
    case 'RecValueDeclaration': {
      const res = infer(node.init, env)
      return {
        type: res.type,
        constraints: [],
        env: bind(node.pat, res.type, res.env)
      }
    }
    case 'ConditionalExpression': {
      const res1 = infer(node.pred, env)
      const res2 = infer(node.cons, env)
      const res3 = infer(node.alt, env)
      return {
        type: node.smlType,
        constraints: [
          ...res1.constraints,
          ...res2.constraints,
          ...res3.constraints,
          new Constraint(res1.type, boolType(), node),
          new Constraint(node.smlType, res2.type, node),
          new Constraint(res2.type, res3.type, node),
        ],
        env
      }
    }
    case 'ApplicationExpression': {
      const resCallee = infer(node.callee, env)
      const resArgs = infer(node.args, env)
      const fnType = functionType(resArgs.type, node.smlType)

      if (resCallee.type.type !== 'function') {
        throw new TypeInferenceError('Function application on non-callable expression', node)
      }

      if (resCallee.type.paramType.type === 'tuple' && fnType.paramType.type === 'tuple') {
        resCallee.type.paramType.elementTypes.reverse()
        fnType.paramType.elementTypes.reverse()
      }

      const constraints = [
        ...resCallee.constraints,
        ...resArgs.constraints,
        new Constraint(resCallee.type, fnType, node),
      ]

      return {
        type: node.smlType,
        constraints,
        env
      }
    }
    case 'ListExpression': {
      const inferredElements = node.elements.map(elem => infer(elem, env))
      let constraints = inferredElements.flatMap(res => res.constraints)

      if (inferredElements.length !== 0) {
        for (let i = 0; i < inferredElements.length - 1; i++) {
          constraints.push(
            new Constraint(inferredElements[i].type, inferredElements[i + 1].type, node)
          )
        }
        constraints.push(new Constraint(node.smlType, listType(inferredElements[0].type), node))
      }
      return {
        type: node.smlType,
        constraints,
        env
      }
    }
    case 'TupleExpression': {
      const inferredElements = node.elements.map(elem => infer(elem, env))
      const inferredType = tupleType(inferredElements.map(elem => elem.type))

      return {
        type: node.smlType,
        constraints: [
          ...inferredElements.flatMap(elem => elem.constraints),
          new Constraint(node.smlType, inferredType, node),
        ],
        env
      }
    }
    case 'Identifier': {
      let type = node.smlType
      if (env.get(node.name) !== undefined) {
        type = env.get(node.name)!
      }
      return {
        type,
        constraints: [],
        env
      }
    }
    case 'Literal':
      return {
        type: node.smlType,
        constraints: [],
        env
      }
    case 'FunctionDeclaration':
    case 'LocalDeclaration':
    case 'DeclarationList':
    case 'LetExpression':
    case 'SequenceExpression':
    case 'LambdaExpression':
      throw new TypeInferenceError(`Not supported yet: ${node.type}`)
    default:
      throw new TypeInferenceError(`Unknown node type: ${node.type}`)
  }
}

class Constraint {
  srcNode: es.Node
  from: es.Type
  to: es.Type
  constructor(from: es.Type, to: es.Type, srcNode: es.Node) {
    this.from = from
    this.to = to
  }
}

class Substitution {
 
  constructor(public  subs: Map<number, es.Type> = new Map()) {}

  insert(before: es.VariableType, after: es.Type): Substitution  {
    const id = before.id
    if (this.subs.has(id) && !isTypeEqual(this.subs.get(id)!, after)) {
      throw new TypeInferenceError('Clashing types found!')
    }
    this.subs.set(id, after)
    return this
  }
  concat(other: Substitution): Substitution  {
    for (const [id, type] of other.subs.entries()) {
      if (this.subs.has(id) && !isTypeEqual(this.subs.get(id)!, type)) {
        throw new TypeInferenceError('Clashing types found!')
      }
      this.subs.set(id, type)
    }
    return this
  }
}

/**
 * Combines two substitutions s1 and s2, and returns s3 = s2.s1, where
 * s3(x) = s2(s1(x)). i.e. s1 is applied before s2 is applied.
 */
function combine(s1: Substitution, s2: Substitution): Substitution {
  console.log("combining")
  console.log(cloneDeep(s1))
  console.log(cloneDeep(s2))
  let s3 = cloneDeep(s1)
  // apply s2 on the RHS of s1
  for (const [id, t] of s3.subs.entries()) {
    s3.subs.set(id, substitute(t, s2))
  }
  // add remaining s2 mappings that are not in s1
  for (const [id, t] of s2.subs.entries()) {
    if (!s3.subs.has(id)) {
      s3.subs.set(id, t)
    }
  }
  console.log(cloneDeep(s3))
  return s3
}

/**
 * Checks if a type contains another type
 */
function contains(t2: es.Type, t1: es.VariableType): boolean {
  if (t2.type === 'variable') {
    return t2.id == t1.id
  } else if (t2.type === 'function') {
    return contains(t2.paramType, t1) || contains(t2.returnType, t1)
  } else if (t2.type === 'list') {
    return contains(t2.elementType, t1)
  } else if (t2.type === 'tuple') {
    for (let t of t2.elementTypes) {
      if (contains(t, t1)) return true
    }
    return false
  } else {
    return false
  }
}

/**
 * Unify a list of constraints
 */
function unify(C: Constraint[]): Substitution {
  console.log("unify", C)
  if (C.length === 0) {
    // If C is the empty set, then unify(C) is the empty substitution.
    return new Substitution()
  } else {
    // If C contains at least one constraint t1 = t2
    // and possibly some other constraints C'.

    const constraint = C[0]
    const t1 = constraint.from
    const t2 = constraint.to

    // If t1 and t2 are both the same simple type — i.e. both the same type
    // variable 'x, or both int or both bool — then return unify(C'). In this
    // case, the constraint contained no useful information, so we’re tossing
    // it out and continuing.
    if (isTypeEqual(t1, t2)) {
      return unify(C.slice(1))
    }

    // If t1 is a type variable 'x and 'x does not occur in t2, then
    // let S = {t2 / 'x}, and return S; unify(C' S). In this case, we are
    // eliminating the variable 'x from the system of equations, much like
    // Gaussian elimination in solving algebraic equations.
    else if (t1.type === 'variable' && !contains(t2, t1)) {
      return combine(new Substitution().insert(t1, t2), unify(C.slice(1)))
    }

    // previous case but swap
    else if (t2.type === 'variable' && !contains(t1, t2)) {
      return combine(new Substitution().insert(t2, t1), unify(C.slice(1)))
    }

    // both are function types
    else if (t1.type === 'function' && t2.type === 'function') {
      return unify(
        [
          new Constraint(t1.returnType, t2.returnType, constraint.srcNode),
          new Constraint(t1.paramType, t2.paramType, constraint.srcNode),
          ...C.slice(1),
        ]
      )
    }

    // both are list types
    else if (t1.type === 'list' && t2.type === 'list') {
      return unify(
        [
          new Constraint(t1.elementType, t2.elementType, constraint.srcNode),
          ...C.slice(1),
        ]
      )
    }

    // both are tuple types
    else if (t1.type === 'tuple' && t2.type === 'tuple') {
      if (t1.elementTypes.length !== t2.elementTypes.length) {
        throw new TypeInferenceError(
          'Type inference error: attempting to match tuples with differing length',
          constraint.srcNode
        )
      }

      const constraints: Constraint[] = []
      for (let i = 0; i < t1.elementTypes.length; i++) {
        constraints.push(
          new Constraint(t1.elementTypes[i], t2.elementTypes[i], constraint.srcNode)
        )
      }

      return unify([...constraints, ...C.slice(1)])
    } else {
      throw new TypeInferenceError('Type inference error: unification failed', constraint.srcNode)
    }
  }
}

function substitute(t: es.Type, S: Substitution): es.Type {
  if (t.type === 'variable') {
    for (const [id, type] of S.subs.entries()) {
      if (t.id === id) {
        return type
      }
    }
    return t // no valid substitution found, return variable type first
  } else if (t.type === 'function') {
    return {
      type: 'function',
      paramType: substitute(t.paramType, S),
      returnType: substitute(t.returnType, S)
    }
  } else if (t.type === 'list') {
    return {
      type: 'list',
      elementType: substitute(t.elementType, S)
    }
  } else if (t.type === 'tuple') {
    return {
      type: 'tuple',
      elementTypes: t.elementTypes.map(typ => substitute(typ, S))
    }
  } else {
    return t
  }
}

const GlobalEnv = new TypeEnv({ ...builtinFunctionTypes })

const runningEnv = new TypeEnv({ ...GlobalEnv.map })

export function inferProgram(node: es.Program, context: Context) {
  try {
    infer(node, runningEnv)
  } catch (error) {
    if (error instanceof TypeError) {
      context.errors.push(error)
    } else {
      throw error
    }
  }
}
