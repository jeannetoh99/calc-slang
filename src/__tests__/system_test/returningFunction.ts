import { DecResType, formatResults } from '../../utils/stringify'
import { expectResult } from '../../utils/testing'

const program = `
    val a = (fn x => fn y => fn z => x + y * z);
    a 2;
    a 1;
    [a 1 2 3, 1];
`

test('returning function', () => {
  const expected: DecResType[] = [
    {
      name: 'a',
      value: 'fn',
      type: 'int -> int -> int -> int'
    },
    {
      name: 'it',
      value: 'fn',
      type: 'int -> int -> int'
    },
    {
      name: 'it',
      value: 'fn',
      type: 'int -> int -> int'
    },
    {
      name: 'it',
      value: '[7,1]',
      type: 'int list'
    }
  ]
  return expectResult(program).toBe(formatResults(expected))
})
