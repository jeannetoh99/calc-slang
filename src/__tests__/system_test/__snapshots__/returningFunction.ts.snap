// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returning function: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "
    val a = (fn x => fn y => fn z => x + y * z);
    a 2;
    a 1;
    [a 1 2 3, 1];
",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "val a = fn : int -> int -> int -> int;
val it = fn : int -> int -> int;
val it = fn : int -> int -> int;
val it = [7,1] : int list;",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
